pkg/common-controller/snapshot_controller.go

```diff
@@ -930,19 +931,22 @@ func (ctrl *csiSnapshotCommonController) ensurePVCFinalizer(snapshot *crdv1.Volu
 
 // removePVCFinalizer removes a Finalizer for VolumeSnapshot Source PVC.
 func (ctrl *csiSnapshotCommonController) removePVCFinalizer(pvc *v1.PersistentVolumeClaim) error {
-	// Get snapshot source which is a PVC
-	// TODO(xyang): We get PVC from informer but it may be outdated
-	// Should get it from API server directly before removing finalizer
-	pvcClone := pvc.DeepCopy()
-	pvcClone.ObjectMeta.Finalizers = utils.RemoveString(pvcClone.ObjectMeta.Finalizers, utils.PVCFinalizer)
-
-	_, err := ctrl.client.CoreV1().PersistentVolumeClaims(pvcClone.Namespace).Update(context.TODO(), pvcClone, metav1.UpdateOptions{})
-	if err != nil {
-		return newControllerUpdateError(pvcClone.Name, err.Error())
-	}
+	return retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		// Get snapshot source which is a PVC
+		newPvc, err := ctrl.client.CoreV1().PersistentVolumeClaims(pvc.Namespace).Get(context.TODO(), pvc.Name, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+		newPvc = newPvc.DeepCopy()
+		newPvc.ObjectMeta.Finalizers = utils.RemoveString(newPvc.ObjectMeta.Finalizers, utils.PVCFinalizer)
+		_, err = ctrl.client.CoreV1().PersistentVolumeClaims(newPvc.Namespace).Update(context.TODO(), newPvc, metav1.UpdateOptions{})
```

Comment:
Using a "patch" instead of an "update" can reduce the conflicts.

Reply:
I do agree that a using a "patch" would get rid of all the conflicts however since the `finalizers` are stored in a list, using a patch seems dangerous to me because if the finalizer list changes between the get and update calls (because another controller removed its finalizer for instance), it means that we would end up removing the wrong finalizer (since we need the index to remove an element from a list using a patch).

Is there something I'm missing that could let us use a patch instead without the drawback above?